servicetalk-concurrent-api-test/src/test/java/io/servicetalk/concurrent/api/test/PublisherStepVerifierTest.java

```diff
@@ -0,0 +1,649 @@
+                    AsyncContext.put(ASYNC_KEY, 15);
+                })
+                .expectError(error -> {
+                    assertSame(DELIBERATE_EXCEPTION, error);
+                    assertThat(AsyncContext.get(ASYNC_KEY), is(15));
+                })
+                .verify();
+    }
+
+    @Test
+    public void thenAwaitExitsWhenVerifyComplete() {
+        create(from("foo"))
+                .thenAwait(ofDays(1))
+                .expectNext("foo")
+                .expectComplete()
+                .verify();
+    }
+
+    @Test
+    public void thenAwaitRespectsDelaysComplete() {
+        PublisherSource.Processor<String, String> processor = newPublisherProcessor();
+        new InlinePublisherFirstStep<>(processor, new DefaultModifiableTimeSource())
+                .expectSubscription(s -> { })
+                .expectNoSignals(ofDays(500))
+                .thenAwait(ofDays(1000))
+                .then(() -> processor.onNext("foo"))
+                .expectNext("foo")
+                .then(processor::onComplete)
+                .expectComplete()
+                .verify();
+    }
+
+    @Test(expected = AssertionError.class)
+    public void thenAwaitRespectsDelaysEqualsFail() {
+        thenAwaitRespectsDelaysFail(true);
+    }
+
+    @Test(expected = AssertionError.class)
+    public void thenAwaitRespectsDelaysGTFail() {
+        thenAwaitRespectsDelaysFail(false);
+    }
+
+    private static void thenAwaitRespectsDelaysFail(boolean equals) {
+        PublisherSource.Processor<String, String> processor = newPublisherProcessor();
+        verifyException(() -> new InlinePublisherFirstStep<>(processor, new DefaultModifiableTimeSource())
+                .expectSubscription(s -> { })
+                .expectNoSignals(ofDays(equals ? 1000 : 1001))
+                .thenAwait(ofDays(1000))
+                .then(() -> processor.onNext("foo"))
+                .expectNext("foo")
+                .then(processor::onComplete)
+                .expectComplete()
+                .verify());
+    }
+
+    private static void verifyException(Supplier<Duration> verifier) {
+        verifyException(verifier, PublisherStepVerifierTest.class.getName());
+    }
+
+    static void verifyException(Supplier<Duration> verifier, String classNamePrefix) {
+        try {
+            verifier.get();
+        } catch (AssertionError error) {
+            StackTraceElement[] stackTraceElements = error.getStackTrace();
+            if (stackTraceElements.length == 0 || !stackTraceElements[0].getClassName().startsWith(classNamePrefix)) {
+                // the tests expect AssertionError, we need to throw a different type of exception!
+                throw new IllegalStateException("stacktrace does not start with classNamePrefix: " + classNamePrefix,
+                        error);
+            }
+            throw error;
+        }
+        throw new IllegalStateException("expected test to fail, but it didn't");
```

Comment:
I'll clarify the comment but we can't use `assert*` methods bcz the tests that fail expect `AssertionError` which means they would pass even if these checks failed.

```
// The tests expect AssertionErrors, we need to throw a different type of exception so we use
// IllegalStateException below.
```

Reply:
I will adjust so that these tests are verifying the actual method that is expected to be thrown, and this will change the exception verification. I'm not sure if `assertThrows` can be used though, as it seems limited to exception class type.